Pre-Python 3 unit test roundup

StephenChan opened the pull request:
Additional unit tests before working on Python 3 compatibility.

This particular PR will make very minimal changes to non-test code, so that it's relatively safe to merge. Once this is done, I will start the Python 3 feature branch as discussed in issue #284.

For this PR, I'm thinking of aiming for:

- Issue #54; should already be covered by a commit here
- Issue #77
- Issue #157, since it's close to done
- Parts of issue #155, more focusing on common functionality than edge cases
- Other common functionality that's lacking tests according to `coverage`

beijbom commented:
@StephenChan : really nice to see this WIP PR. I think it sounds like a great effort.

StephenChan commented:
All right, I think I'm ready to move on from testing, unless I messed something up.

- #54, #155, and #157 have been dealt with. (Also #77 in the earlier security PR.) I moved the skipped items to other lower-priority issues.
- Most of the TODO comments in the test .py files have been addressed. Again, I skipped some.
- Two non-test behaviors were changed:
  - Filename generation for uploaded images has been changed a bit. It turns out that this process was already guarded from generating duplicates, but it would accept instances of one image being named `1234567890.png` and another image being named `1234567890.jpg`, which could be confusing, even if not actually dangerous. I changed the code to minimize the possibility of this happening, and if it does ever happen, we'll get an email. See `images/models.py`.
  - The `label_new_ajax` view now responds with JSON instead of HTML when the user isn't logged in. See `labels/views.py`. This is a corner case since you have to be on the labelset add/remove page to submit to this view normally, and that page requires login.
- Code coverage is up from 91% to 93%. Not that big of a jump, but it's something. Most of the remaining un-covered lines fall into a few tricky categories:
  - Backend tasks and functions related to backend_overview
  - Statistics page (it's still broken, of course)
  - Selenium and other tests that are skipped for now (yes, tests themselves are counted in the coverage)
  - Migration if/else cases which check for existing data (we'll squash most of these files sooner or later anyway)

StephenChan commented:
...As I'm starting to look at Python 3 porting now, I should note: I may still add more tests for things that come to mind during Python 3 porting. But this particular PR has enough stuff in it.
